src/execution_plan/execution_plan.c

```diff
@@ -498,304 +482,360 @@ static void _NewExecutionPlanSegment(RedisModuleCtx *ctx, GraphContext *gc,
-		const cypher_astnode_t *skip_clause = cypher_ast_return_get_skip(ret_clause);
-		const cypher_astnode_t *limit_clause = cypher_ast_return_get_limit(ret_clause);
-
-		uint skip = 0;
-		uint limit = 0;
-		if(skip_clause) skip = AST_ParseIntegerNode(skip_clause);
-		if(limit_clause) limit = AST_ParseIntegerNode(limit_clause);
-
-		if(order_expressions) {
-			int direction = AST_PrepareSortOp(order_clause);
-			// The sort operation will obey a specified limit, but must account for skipped records
-			uint sort_limit = (limit > 0) ? limit + skip : 0;
-			op = NewSortOp(order_expressions, direction, sort_limit);
-			Vector_Push(ops, op);
-		}
+	// Build filter tree
+	FT_FilterNode *filter_tree = AST_BuildFilterTree(ast, record_map);
+	segment->filter_tree = filter_tree;
 
-		if(skip_clause) {
-			OpBase *op_skip = NewSkipOp(skip);
-			Vector_Push(ops, op_skip);
-		}
+	// Instantiate the Vector that will hold all constructed operations.
+	// TODO This vector could be removed with some logical changes.
+	uint clause_count = cypher_ast_query_nclauses(ast->root);
+	Vector *ops = NewVector(OpBase *, clause_count);
 
-		if(limit_clause) {
-			OpBase *op_limit = NewLimitOp(limit);
-			Vector_Push(ops, op_limit);
-		}
+	/* If we have a procedure call, build an op for it first.
+	 * This is currently necessary to extend the segment's projections if necesary
+	 * and in order to build the plan for queries of forms like CALL..MATCH properly.
+	 * TODO Fix this so that CALL clauses are handled in the ConvertClause loop. */
+	const cypher_astnode_t *call_clause = AST_GetClause(ast, CYPHER_AST_CALL);
+	if(call_clause) {
+		OpBase *opProcCall = _buildCallOp(ast, segment, call_clause);
+		Vector_Push(ops, opProcCall);
+	}
 
-		op = NewResultsOp(result_set, qg);
-		Vector_Push(ops, op);
-	} else if(call_clause) {
-		op = NewResultsOp(result_set, qg);
-		Vector_Push(ops, op);
+	// Build traversal operations for every connected component in the QueryGraph
+	if(AST_ContainsClause(ast, CYPHER_AST_MATCH) || AST_ContainsClause(ast, CYPHER_AST_MERGE)) {
+		_ExecutionPlanSegment_ProcessQueryGraph(segment, qg, ast, filter_tree, ops);
+	}
+
+	// If we are in a querying context, retrieve a pointer to the statistics for operations
+	// like DELETE that only produce metadata.
+	ResultSetStatistics *stats = (result_set) ? &result_set->stats : NULL;
+
+	for(uint i = 0; i < clause_count; i ++) {
+		// Build the appropriate operation(s) for each clause in the query.
+		const cypher_astnode_t *clause = cypher_ast_query_get_clause(ast->root, i);
+		_ExecutionPlanSegment_ConvertClause(gc, ast, segment, stats, clause, ops);
+	}
+
+	// The root operation is OpResults only if the query culminates in a RETURN or CALL clause.
+	cypher_astnode_type_t last_clause_type = cypher_astnode_type(cypher_ast_query_get_clause(ast->root,
+																 clause_count - 1));
+	if(last_clause_type == CYPHER_AST_RETURN || last_clause_type == CYPHER_AST_CALL) {
```

Comment:
What about `CALL proc() YIELD x MATCH (a)-[]->(x) WITH a AS q`
in this case we don't want to introduce a `ResultOp`.

Reply:
On master this is fine, since the final projection is part of the same segment. On `hashable-records`, we'll instead check to see if we're building the last segment (or possibly inject the ResultsOp outside of segment construction).